### DISCUSSION TITLE
How do I modify a node module and use that in Vercel?

### DISCUSSION THREAD

**Discussion by shajidhasan** [vercel/vercel#6391] [Help]
During every build, Vercel downloads all the dependencies so I can't just edit a local file in my node_modules and expect that to have any effect in Vercel. I have tried `patch-package` as a `postinstall` script but patch-package doesn't work during a Vercel build. So how to go around this problem?

**Comment by barelyhuman** [comment]
1. easiest one, fork the original package, modify source, use your git repo as the dependency in package.json.
2. ruins the ease but instead of using the vercel web , use the cli command to run the setup instead and the patch-package should work fine here. 
3. Use a simple shell script to make the needed changes before the actual build. 

3rd one works because the build step is run after the environment is setup up which is where the installation of the packages go through and you can just change you build step in Project Settings > General , to run the shell script, before it runs the build command.

examples: `npx patch-package; npm run build` or `./patch.sh; npm run build` 

let me know if that works

**Reply by Bhaney44** [reply]
I'm having a similar issue. Vercel is awesome - one of the best technologies I have ever used. But, I need to make one edit to one file in my `node_modules` to avoid breaking problem during build. What is the best way to do that? I don't understand how I would modify my source because I don't have any `node_modules` in my source.

**Reply by leerob** [reply]
`patch-package` should definitely work, you'll need to check in those changes to git. I've used this before on my site!

**Reply by Bhaney44** [reply]
Thanks @leerob! Is there any documentation on `patch-package`, or do you have suggestions on where I can find information on how to use it in Vercel?

**Reply by Bhaney44** [reply]
Found the [docs](https://www.npmjs.com/package/patch-package)!

**Reply by barelyhuman** [reply]
Also make sure you pin the dependency to the exact version when using patch-package to avoid issues with patched updates. 
Though `patch-package` already does that while it tries to install but you can avoid the extra check and install by doing that yourself.

**Reply by Bhaney44** [reply]
I ended up trying the first option to fork the repo I needed. I then updated my `package.json` file accordingly. The fork method is working locally, but I am still getting an error during the build.

```
[2/4] Fetching packages...
--
12:33:06.442 | error Command failed.
12:33:06.442 | Exit code: 128
12:33:06.442 | Command: git
12:33:06.442 | Arguments: ls-remote --tags --heads git://github.com/ChoiceCoin/react-scripts.git
12:33:06.442 | Directory: /vercel/path0
12:33:06.442 | Output:
12:33:06.443 | fatal: unable to connect to github.com:
12:33:06.443 | github.com[0: 192.30.255.113]: errno=Connection timed out
12:33:06.443 | info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
12:33:06.463 | Running "yarn run build"
12:33:06.687 | yarn run v1.22.17
12:33:06.733 | $ react-scripts build
12:33:06.750 | /bin/sh: react-scripts: command not found
12:33:06.765 | error Command failed with exit code 127.
12:33:06.765 | info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
12:33:06.773 | Error: Command "yarn run build" exited with 127
```
I'm not sure why I am receiving this error. I'm thinking it may be a problem with the formatting of the link in the `package.json` file.

`"react-scripts": "git://github.com/ChoiceCoin/react-scripts.git",` (works locally)(failed via Vercel)

Trying new formats:
`"react-scripts": "git:https://github.com/ChoiceCoin/react-scripts",` (failed)
`"react-scripts": "git:ChoiceCoin/react-scripts",` (failed)

The fatal error I don't understand is:

`12:33:06.443 | fatal: unable to connect to github.com:`

Trying overrides in Vercel `Project Settings`:

`npm install` (failed)
`yarn install` (failed)
`npm run build ` (failed)
`react-scripts build` (failed)

Any suggestions or advice would be greatly appreciated. Thanks! @barelyhuman @leerob

**Reply by amyegan** [reply]
Hi @Bhaney44! Is the package published publicly, or is it a private package/repo? If it is private then you will need to provide a token to allow it to install. The guide [How do I use private dependencies with Vercel?](https://vercel.com/guides/using-private-dependencies-with-vercel) has more information. GitHub Packages documentation on [working with the npm registry](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-npm-registry) might also be helpful.

**Reply by Bhaney44** [reply]
Hi @amyegan - thanks for the reply! Thanks for the helpful documentation. The GitHub was public, so I'm not sure why there would be an issue. I ended up solving this issue by changing the dependencies, which allowed me to avoid the need to use a fix in node modules. But, this is helpful because I am sure I will run into the problem again.

**Reply by amyegan** [reply]
I'm glad you found a solution that works for your app!

**Reply by barelyhuman** [reply]
Thanks @amyegan , 
apparently mail notifications are off

@Bhaney44 If you could let me know what you were using as a format for the package dep that might be the issue. 

I normally use `git+ssh://git@github.com/<username>/<repo>.git` and while testing locally make sure the installed deb does actually build itself if necessary.

Also, yarn, npm and pnpm have a slightly different approach at git deps so you might wanna stick to a single dep manager for this and only it's lock file. 


Hopefully that helps in the future, if you wish to try something like this again

**Reply by Bhaney44** [reply]
> Thanks @amyegan , apparently mail notifications are off
> 
> @Bhaney44 If you could let me know what you were using as a format for the package dep that might be the issue.
> 
> I normally use `git+ssh://git@github.com/<username>/<repo>.git` and while testing locally make sure the installed deb does actually build itself if necessary.
> 
> Also, yarn, npm and pnpm have a slightly different approach at git deps so you might wanna stick to a single dep manager for this and only it's lock file.
> 
> Hopefully that helps in the future, if you wish to try something like this again

Thanks for your help. I think you're right - it was my formatting.

```
"react-scripts": "git://github.com/ChoiceCoin/react-scripts.git", (works locally)(failed via Vercel)
"react-scripts": "git:https://github.com/ChoiceCoin/react-scripts", (failed)
"react-scripts": "git:ChoiceCoin/react-scripts", (failed)
```

In the future, I will run: 

`git+ssh://git@github.com/<username>/<repo>.git`

Thanks!

**Comment by beadlespouse** [comment]
Looks like patch-package is now working on vercel! Much easier than forking...

**Reply by timtjoe** [reply]
sure!
